packages/components/psammead-content-shift-blocker/src/index.jsx

```diff
@@ -0,0 +1,160 @@
+    /* webpackChunkName: "intersection-observer-polyfill" */ 'intersection-observer'
+  ).then(() => {
+    IntersectionObserver.prototype.POLL_INTERVAL = 100;
+    return init();
+  });
+};
+
+const initResizeObserver = ({ wrapperEl, setContentElRect }) => {
+  const init = ResizeObserver => {
+    const callback = ([contentEntry]) => {
+      setContentElRect(contentEntry.contentRect);
+    };
+    const RO = new ResizeObserver(callback);
+
+    RO.observe(wrapperEl.current.firstChild);
+
+    return RO;
+  };
+
+  if ('ResizeObserver' in window) {
+    return init(ResizeObserver);
+  }
+  return import(
+    /* webpackChunkName: "resize-observer-polyfill" */ '@juggle/resize-observer'
+  ).then(module => {
+    const ResizeObserver = module.default;
+    return init(ResizeObserver);
+  });
+};
+
+const isScrollAnchoringSupported = () => {
+  // https://drafts.csswg.org/css-scroll-anchoring/
+  if ('CSS' in window) {
+    return CSS.supports('overflow-anchor', 'auto');
+  }
+  return false;
+};
+
+const ContentShiftBlocker = ({ children, initialHeight, initialWidth }) => {
+  const wrapperEl = useRef(null);
+  const scrollHeight = useRef(null);
+  const scrollAnchoringIsSupported = useRef(null);
+  const [wrapperIO, setWrapperIO] = useState({});
+  const [contentElRect, setContentElRect] = useState({});
+  const [wrapperDimensions, setWrapperDimensions] = useState({
+    height: initialHeight,
+    width: initialWidth,
+  });
+
+  useEffect(() => {
+    // component did mount
+    const IO = initIntersectionObserver({ wrapperEl, setWrapperIO });
+    const RO = initResizeObserver({ wrapperEl, setContentElRect });
+    scrollAnchoringIsSupported.current = isScrollAnchoringSupported();
+
+    return function cleanup() {
+      IO.disconnect();
+      RO.disconnect();
+    };
+  }, []);
+
+  useLayoutEffect(() => {
+    // child content did resize
+    const wrapperIsOutOfView = wrapperIO.isIntersecting === false;
+
+    if (wrapperIsOutOfView) {
+      // wrapper will resize
+      const wrapperIsAboveViewport = wrapperIO.boundingClientRect.top < 0;
+      const {
+        width: nextWrapperWidth,
+        height: nextWrapperHeight,
+      } = contentElRect;
+
+      if (!scrollAnchoringIsSupported.current && wrapperIsAboveViewport) {
+        scrollHeight.current = document.body.scrollHeight;
+      }
+
+      setWrapperDimensions({
+        width: nextWrapperWidth,
+        height: nextWrapperHeight,
+      });
+    }
+  }, [contentElRect]);
+
+  useEffect(() => {
+    // wrapper did resize
+    if (scrollAnchoringIsSupported.current || scrollHeight.current === null) {
+      return;
+    }
+
+    const { scrollHeight: newScrollHeight } = document.body;
+    const { current: prevScrollHeight } = scrollHeight;
+    const scrollHeightDiff = prevScrollHeight - newScrollHeight;
+
+    if (scrollHeightDiff !== 0) {
+      // adjust scrollY position to prevent visible jump
+      window.scrollTo(0, window.pageYOffset - scrollHeightDiff);
+    }
```

Comment:
@andrewscfc I'm not sure how to make this a pure function when it relies heavily on state, refs and the window object outside of it's own scope. Maybe we need to have a chat about what you're thinking.

Reply:
@jroebu14, here's what I was thinking:

```javascript
    const { scrollHeight: newScrollHeight } = document.body;
    const { current: prevScrollHeight } = scrollHeight;
    const { pageYOffset : scrollY } = window;

    const scrollHeight = calculateScrollHeight({ newScrollHeight, prevScrollHeight, scrollY })
    // adjust scrollY position to prevent visible jump
    if(scrollHeight) {
      window.scrollTo(0, scrollHeight);
    }

    const calculateScrollHeight = ({ newScrollHeight, prevScrollHeight, scrollY }) => {
      scrollHeightDiff = prevScrollHeight - newScrollHeight;
      
      if(scrollHeightDiff !== 0) {
        return scrollY - scrollHeightDiff;
      }
      return null;
    }
```
The pure function is `calculateScrollHeight()` but maybe it's a bit to simple to merit its own module, when I looked over your tests carefully I was more comfortable: https://github.com/bbc/psammead/pull/2582/files#r350880186

Reply:
Nice idea. I've updated the code. Thanks!